File: github/typescript/src/services/repositories.ts
/**
 * GitHub Repositories Service
 *
 * Provides comprehensive repository management functionality including:
 * - Repository CRUD operations
 * - Branch management
 * - File content operations
 * - Release management
 *
 * @module services/repositories
 */

/**
 * Parameters for listing repositories
 */
export interface ListReposParams {
  /** Filter by visibility: public, private, or all */
  visibility?: 'public' | 'private' | 'all';
  /** Comma-separated list of values: owner, collaborator, organization_member */
  affiliation?: string;
  /** Filter by type: all, owner, public, private, member */
  type?: 'all' | 'owner' | 'public' | 'private' | 'member';
  /** Sort by: created, updated, pushed, full_name */
  sort?: 'created' | 'updated' | 'pushed' | 'full_name';
  /** Sort direction: asc or desc */
  direction?: 'asc' | 'desc';
  /** Results per page (max 100) */
  per_page?: number;
  /** Page number */
  page?: number;
}

/**
 * Request to create a new repository
 */
export interface CreateRepoRequest {
  /** Repository name */
  name: string;
  /** Repository description */
  description?: string;
  /** Homepage URL */
  homepage?: string;
  /** Make repository private */
  private?: boolean;
  /** Visibility: public, private, or internal */
  visibility?: 'public' | 'private' | 'internal';
  /** Enable issues */
  has_issues?: boolean;
  /** Enable projects */
  has_projects?: boolean;
  /** Enable wiki */
  has_wiki?: boolean;
  /** Enable downloads */
  has_downloads?: boolean;
  /** Make this a template repository */
  is_template?: boolean;
  /** Team ID with access (orgs only) */
  team_id?: number;
  /** Initialize with README */
  auto_init?: boolean;
  /** .gitignore template name */
  gitignore_template?: string;
  /** License template name */
  license_template?: string;
  /** Allow squash merge */
  allow_squash_merge?: boolean;
  /** Allow merge commits */
  allow_merge_commit?: boolean;
  /** Allow rebase merge */
  allow_rebase_merge?: boolean;
  /** Allow auto-merge */
  allow_auto_merge?: boolean;
  /** Delete branch on merge */
  delete_branch_on_merge?: boolean;
}

/**
 * Request to update a repository
 */
export interface UpdateRepoRequest {
  /** Repository name */
  name?: string;
  /** Repository description */
  description?: string;
  /** Homepage URL */
  homepage?: string;
  /** Make repository private */
  private?: boolean;
  /** Visibility: public, private, or internal */
  visibility?: 'public' | 'private' | 'internal';
  /** Enable issues */
  has_issues?: boolean;
  /** Enable projects */
  has_projects?: boolean;
  /** Enable wiki */
  has_wiki?: boolean;
  /** Default branch */
  default_branch?: string;
  /** Allow squash merge */
  allow_squash_merge?: boolean;
  /** Allow merge commits */
  allow_merge_commit?: boolean;
  /** Allow rebase merge */
  allow_rebase_merge?: boolean;
  /** Allow auto-merge */
  allow_auto_merge?: boolean;
  /** Delete branch on merge */
  delete_branch_on_merge?: boolean;
  /** Archive repository */
  archived?: boolean;
}

/**
 * Request to create or update a file
 */
export interface CreateOrUpdateFileRequest {
  /** Commit message */
  message: string;
  /** File content (base64 encoded for binary files) */
  content: string;
  /** SHA of file being replaced (required for updates) */
  sha?: string;
  /** Branch name (defaults to default branch) */
  branch?: string;
  /** Committer details */
  committer?: {
    name: string;
    email: string;
  };
  /** Author details */
  author?: {
    name: string;
    email: string;
  };
}

/**
 * Request to create a release
 */
export interface CreateReleaseRequest {
  /** Tag name */
  tag_name: string;
  /** Target commitish (branch or commit SHA) */
  target_commitish?: string;
  /** Release name */
  name?: string;
  /** Release description */
  body?: string;
  /** Mark as draft */
  draft?: boolean;
  /** Mark as prerelease */
  prerelease?: boolean;
  /** Generate release notes automatically */
  generate_release_notes?: boolean;
}

/**
 * Request to update a release
 */
export interface UpdateReleaseRequest {
  /** Tag name */
  tag_name?: string;
  /** Target commitish */
  target_commitish?: string;
  /** Release name */
  name?: string;
  /** Release description */
  body?: string;
  /** Mark as draft */
  draft?: boolean;
  /** Mark as prerelease */
  prerelease?: boolean;
}

/**
 * Repository representation
 */
export interface Repository {
  id: number;
  node_id: string;
  name: string;
  full_name: string;
  owner: any; // User type
  private: boolean;
  html_url: string;
  description: string | null;
  fork: boolean;
  url: string;
  created_at: string;
  updated_at: string;
  pushed_at: string;
  git_url: string;
  ssh_url: string;
  clone_url: string;
  language: string | null;
  forks_count: number;
  stargazers_count: number;
  watchers_count: number;
  size: number;
  default_branch: string;
  open_issues_count: number;
  topics: string[];
  has_issues: boolean;
  has_projects: boolean;
  has_wiki: boolean;
  has_downloads: boolean;
  archived: boolean;
  disabled: boolean;
  visibility: string;
  license: any | null;
}

/**
 * Branch representation
 */
export interface Branch {
  name: string;
  commit: {
    sha: string;
    url: string;
  };
  protected: boolean;
}

/**
 * Content item (file or directory)
 */
export interface ContentItem {
  type: 'file' | 'dir' | 'symlink' | 'submodule';
  size: number;
  name: string;
  path: string;
  content?: string; // Base64 encoded for files
  sha: string;
  url: string;
  git_url: string | null;
  html_url: string | null;
  download_url: string | null;
}

/**
 * Release representation
 */
export interface Release {
  id: number;
  node_id: string;
  tag_name: string;
  target_commitish: string;
  name: string | null;
  body: string | null;
  draft: boolean;
  prerelease: boolean;
  created_at: string;
  published_at: string | null;
  url: string;
  html_url: string;
  assets_url: string;
  upload_url: string;
  tarball_url: string | null;
  zipball_url: string | null;
  assets: any[];
}

/**
 * Paginated response wrapper
 */
export interface Paginated<T> {
  items: T[];
  total_count?: number;
  next_page?: string;
  prev_page?: string;
  first_page?: string;
  last_page?: string;
}

/**
 * Repositories Service
 *
 * Handles all repository-related operations including CRUD, branches,
 * contents, and releases.
 */
export class RepositoriesService {
  /**
   * Creates a new RepositoriesService instance
   *
   * @param client - Reference to the GitHub client for making API calls
   */
  constructor(private client: any) {}

  /**
   * List repositories for the authenticated user
   *
   * @param params - Optional parameters for filtering and pagination
   * @returns Paginated list of repositories
   *
   * @example
   * ```typescript
   * const repos = await client.repositories.listForUser({
   *   visibility: 'public',
   *   sort: 'updated',
   *   per_page: 50
   * });
   * ```
   */
  async listForUser(params?: ListReposParams): Promise<Paginated<Repository>> {
    return this.client.request<Paginated<Repository>>('GET', '/user/repos', {
      params
    });
  }

  /**
   * List repositories for an organization
   *
   * @param org - Organization name
   * @param params - Optional parameters for filtering and pagination
   * @returns Paginated list of repositories
   *
   * @example
   * ```typescript
   * const repos = await client.repositories.listForOrg('myorg', {
   *   type: 'public',
   *   per_page: 100
   * });
   * ```
   */
  async listForOrg(org: string, params?: ListReposParams): Promise<Paginated<Repository>> {
    return this.client.request<Paginated<Repository>>('GET', `/orgs/${org}/repos`, {
      params
    });
  }

  /**
   * List repositories for the authenticated user (legacy method)
   *
   * @param params - Optional parameters for filtering and pagination
   * @returns Paginated list of repositories
   */
  async listForAuthenticatedUser(params?: ListReposParams): Promise<Paginated<Repository>> {
    return this.listForUser(params);
  }

  /**
   * Get a repository by owner and name
   *
   * @param owner - Repository owner (user or organization)
   * @param repo - Repository name
   * @returns Repository details
   *
   * @example
   * ```typescript
   * const repo = await client.repositories.get('octocat', 'hello-world');
   * console.log(repo.description);
   * ```
   */
  async get(owner: string, repo: string): Promise<Repository> {
    return this.client.request<Repository>('GET', `/repos/${owner}/${repo}`);
  }

  /**
   * Create a repository for the authenticated user
   *
   * @param request - Repository creation parameters
   * @returns Created repository
   *
   * @example
   * ```typescript
   * const repo = await client.repositories.create({
   *   name: 'my-new-repo',
   *   description: 'My awesome project',
   *   private: false,
   *   auto_init: true
   * });
   * ```
   */
  async create(request: CreateRepoRequest): Promise<Repository> {
    return this.client.request<Repository>('POST', '/user/repos', {
      body: request
    });
  }

  /**
   * Create a repository in an organization
   *
   * @param org - Organization name
   * @param request - Repository creation parameters
   * @returns Created repository
   *
   * @example
   * ```typescript
   * const repo = await client.repositories.createForOrg('myorg', {
   *   name: 'org-repo',
   *   private: true
   * });
   * ```
   */
  async createForOrg(org: string, request: CreateRepoRequest): Promise<Repository> {
    return this.client.request<Repository>('POST', `/orgs/${org}/repos`, {
      body: request
    });
  }

  /**
   * Update a repository
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @param request - Update parameters
   * @returns Updated repository
   *
   * @example
   * ```typescript
   * const repo = await client.repositories.update('octocat', 'hello-world', {
   *   description: 'Updated description',
   *   has_wiki: false
   * });
   * ```
   */
  async update(owner: string, repo: string, request: UpdateRepoRequest): Promise<Repository> {
    return this.client.request<Repository>('PATCH', `/repos/${owner}/${repo}`, {
      body: request
    });
  }

  /**
   * Delete a repository
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   *
   * @example
   * ```typescript
   * await client.repositories.delete('octocat', 'old-repo');
   * ```
   */
  async delete(owner: string, repo: string): Promise<void> {
    await this.client.request<void>('DELETE', `/repos/${owner}/${repo}`);
  }

  // Branch operations

  /**
   * List branches for a repository
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @returns List of branches
   *
   * @example
   * ```typescript
   * const branches = await client.repositories.listBranches('octocat', 'hello-world');
   * branches.items.forEach(branch => console.log(branch.name));
   * ```
   */
  async listBranches(owner: string, repo: string): Promise<Paginated<Branch>> {
    return this.client.request<Paginated<Branch>>('GET', `/repos/${owner}/${repo}/branches`);
  }

  /**
   * Get a specific branch
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @param branch - Branch name
   * @returns Branch details
   *
   * @example
   * ```typescript
   * const branch = await client.repositories.getBranch('octocat', 'hello-world', 'main');
   * console.log(branch.commit.sha);
   * ```
   */
  async getBranch(owner: string, repo: string, branch: string): Promise<Branch> {
    return this.client.request<Branch>('GET', `/repos/${owner}/${repo}/branches/${branch}`);
  }

  // Content operations

  /**
   * Get repository contents (file or directory)
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @param path - Path to file or directory
   * @param ref - Branch, tag, or commit SHA (optional)
   * @returns Content item(s)
   *
   * @example
   * ```typescript
   * const readme = await client.repositories.getContents('octocat', 'hello-world', 'README.md');
   * const content = Buffer.from(readme.content, 'base64').toString('utf-8');
   * ```
   */
  async getContents(
    owner: string,
    repo: string,
    path: string,
    ref?: string
  ): Promise<ContentItem | ContentItem[]> {
    return this.client.request<ContentItem | ContentItem[]>('GET', `/repos/${owner}/${repo}/contents/${path}`, {
      params: ref ? { ref } : undefined
    });
  }

  /**
   * Create or update a file in a repository
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @param path - Path where to create/update the file
   * @param request - File content and metadata
   * @returns Updated content information
   *
   * @example
   * ```typescript
   * const result = await client.repositories.createOrUpdateFile(
   *   'octocat',
   *   'hello-world',
   *   'docs/new-file.md',
   *   {
   *     message: 'Add new documentation',
   *     content: Buffer.from('# Hello World').toString('base64')
   *   }
   * );
   * ```
   */
  async createOrUpdateFile(
    owner: string,
    repo: string,
    path: string,
    request: CreateOrUpdateFileRequest
  ): Promise<any> {
    return this.client.request<any>('PUT', `/repos/${owner}/${repo}/contents/${path}`, {
      body: request
    });
  }

  /**
   * Delete a file in a repository
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @param path - Path to the file to delete
   * @param message - Commit message
   * @param sha - SHA of the file being deleted
   * @param branch - Branch to delete from (optional)
   *
   * @example
   * ```typescript
   * await client.repositories.deleteFile(
   *   'octocat',
   *   'hello-world',
   *   'old-file.txt',
   *   'Remove obsolete file',
   *   'file-sha-here'
   * );
   * ```
   */
  async deleteFile(
    owner: string,
    repo: string,
    path: string,
    message: string,
    sha: string,
    branch?: string
  ): Promise<void> {
    await this.client.request<void>('DELETE', `/repos/${owner}/${repo}/contents/${path}`, {
      body: {
        message,
        sha,
        branch
      }
    });
  }

  /**
   * Get the README for a repository
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @returns README content
   *
   * @example
   * ```typescript
   * const readme = await client.repositories.getReadme('octocat', 'hello-world');
   * const content = Buffer.from(readme.content, 'base64').toString('utf-8');
   * ```
   */
  async getReadme(owner: string, repo: string): Promise<ContentItem> {
    return this.client.request<ContentItem>('GET', `/repos/${owner}/${repo}/readme`);
  }

  // Release operations

  /**
   * List releases for a repository
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @returns Paginated list of releases
   *
   * @example
   * ```typescript
   * const releases = await client.repositories.listReleases('octocat', 'hello-world');
   * releases.items.forEach(release => console.log(release.tag_name));
   * ```
   */
  async listReleases(owner: string, repo: string): Promise<Paginated<Release>> {
    return this.client.request<Paginated<Release>>('GET', `/repos/${owner}/${repo}/releases`);
  }

  /**
   * Get a specific release by ID
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @param releaseId - Release ID
   * @returns Release details
   *
   * @example
   * ```typescript
   * const release = await client.repositories.getRelease('octocat', 'hello-world', 12345);
   * ```
   */
  async getRelease(owner: string, repo: string, releaseId: number): Promise<Release> {
    return this.client.request<Release>('GET', `/repos/${owner}/${repo}/releases/${releaseId}`);
  }

  /**
   * Get the latest release
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @returns Latest release
   *
   * @example
   * ```typescript
   * const latest = await client.repositories.getLatestRelease('octocat', 'hello-world');
   * console.log(`Latest version: ${latest.tag_name}`);
   * ```
   */
  async getLatestRelease(owner: string, repo: string): Promise<Release> {
    return this.client.request<Release>('GET', `/repos/${owner}/${repo}/releases/latest`);
  }

  /**
   * Get a release by tag name
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @param tag - Tag name
   * @returns Release for the specified tag
   *
   * @example
   * ```typescript
   * const release = await client.repositories.getReleaseByTag('octocat', 'hello-world', 'v1.0.0');
   * ```
   */
  async getReleaseByTag(owner: string, repo: string, tag: string): Promise<Release> {
    return this.client.request<Release>('GET', `/repos/${owner}/${repo}/releases/tags/${tag}`);
  }

  /**
   * Create a new release
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @param request - Release creation parameters
   * @returns Created release
   *
   * @example
   * ```typescript
   * const release = await client.repositories.createRelease('octocat', 'hello-world', {
   *   tag_name: 'v1.0.0',
   *   name: 'Version 1.0.0',
   *   body: 'Initial release',
   *   draft: false,
   *   prerelease: false
   * });
   * ```
   */
  async createRelease(owner: string, repo: string, request: CreateReleaseRequest): Promise<Release> {
    return this.client.request<Release>('POST', `/repos/${owner}/${repo}/releases`, {
      body: request
    });
  }

  /**
   * Update a release
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @param releaseId - Release ID
   * @param request - Update parameters
   * @returns Updated release
   *
   * @example
   * ```typescript
   * const release = await client.repositories.updateRelease('octocat', 'hello-world', 12345, {
   *   name: 'Updated Release Name',
   *   draft: false
   * });
   * ```
   */
  async updateRelease(
    owner: string,
    repo: string,
    releaseId: number,
    request: UpdateReleaseRequest
  ): Promise<Release> {
    return this.client.request<Release>('PATCH', `/repos/${owner}/${repo}/releases/${releaseId}`, {
      body: request
    });
  }

  /**
   * Delete a release
   *
   * @param owner - Repository owner
   * @param repo - Repository name
   * @param releaseId - Release ID
   *
   * @example
   * ```typescript
   * await client.repositories.deleteRelease('octocat', 'hello-world', 12345);
   * ```
   */
  async deleteRelease(owner: string, repo: string, releaseId: number): Promise<void> {
    await this.client.request<void>('DELETE', `/repos/${owner}/${repo}/releases/${releaseId}`);
  }
}

File: github/rust/src/services/repositories.rs
//! Repository operations.

use crate::client::GitHubClient;
use crate::errors::GitHubResult;
use crate::pagination::{Page, PaginationParams};
use crate::types::{Branch, Content, Release, ReleaseAsset, Repository};
use serde::{Deserialize, Serialize};

/// Service for repository operations.
pub struct RepositoriesService<'a> {
    client: &'a GitHubClient,
}

impl<'a> RepositoriesService<'a> {
    /// Creates a new repositories service.
    pub fn new(client: &'a GitHubClient) -> Self {
        Self { client }
    }

    /// Lists repositories for a user.
    pub async fn list_for_user(&self, username: &str) -> GitHubResult<Vec<Repository>> {
        self.list_for_user_paginated(username, &ListReposParams::default())
            .await
    }

    /// Lists repositories for a user with pagination.
    pub async fn list_for_user_paginated(
        &self,
        username: &str,
        params: &ListReposParams,
    ) -> GitHubResult<Vec<Repository>> {
        self.client
            .get_with_params(&format!("/users/{}/repos", username), params)
            .await
    }

    /// Lists repositories for an organization.
    pub async fn list_for_org(&self, org: &str) -> GitHubResult<Vec<Repository>> {
        self.list_for_org_paginated(org, &ListReposParams::default())
            .await
    }

    /// Lists repositories for an organization with pagination.
    pub async fn list_for_org_paginated(
        &self,
        org: &str,
        params: &ListReposParams,
    ) -> GitHubResult<Vec<Repository>> {
        self.client
            .get_with_params(&format!("/orgs/{}/repos", org), params)
            .await
    }

    /// Lists repositories for the authenticated user.
    pub async fn list_for_authenticated_user(&self) -> GitHubResult<Vec<Repository>> {
        self.client.get("/user/repos").await
    }

    /// Gets a repository.
    pub async fn get(&self, owner: &str, repo: &str) -> GitHubResult<Repository> {
        self.client
            .get(&format!("/repos/{}/{}", owner, repo))
            .await
    }

    /// Creates a repository.
    pub async fn create(&self, request: &CreateRepoRequest) -> GitHubResult<Repository> {
        self.client.post("/user/repos", request).await
    }

    /// Creates a repository for an organization.
    pub async fn create_for_org(
        &self,
        org: &str,
        request: &CreateRepoRequest,
    ) -> GitHubResult<Repository> {
        self.client
            .post(&format!("/orgs/{}/repos", org), request)
            .await
    }

    /// Updates a repository.
    pub async fn update(
        &self,
        owner: &str,
        repo: &str,
        request: &UpdateRepoRequest,
    ) -> GitHubResult<Repository> {
        self.client
            .patch(&format!("/repos/{}/{}", owner, repo), request)
            .await
    }

    /// Deletes a repository.
    pub async fn delete(&self, owner: &str, repo: &str) -> GitHubResult<()> {
        self.client
            .delete(&format!("/repos/{}/{}", owner, repo))
            .await
    }

    // Branches

    /// Lists branches in a repository.
    pub async fn list_branches(&self, owner: &str, repo: &str) -> GitHubResult<Vec<Branch>> {
        self.client
            .get(&format!("/repos/{}/{}/branches", owner, repo))
            .await
    }

    /// Gets a branch.
    pub async fn get_branch(&self, owner: &str, repo: &str, branch: &str) -> GitHubResult<Branch> {
        self.client
            .get(&format!("/repos/{}/{}/branches/{}", owner, repo, branch))
            .await
    }

    // Contents

    /// Gets repository contents (file or directory).
    pub async fn get_contents(
        &self,
        owner: &str,
        repo: &str,
        path: &str,
        git_ref: Option<&str>,
    ) -> GitHubResult<Content> {
        let mut url = format!("/repos/{}/{}/contents/{}", owner, repo, path);
        if let Some(r) = git_ref {
            url = format!("{}?ref={}", url, r);
        }
        self.client.get(&url).await
    }

    /// Creates or updates a file.
    pub async fn create_or_update_file(
        &self,
        owner: &str,
        repo: &str,
        path: &str,
        request: &CreateOrUpdateFileRequest,
    ) -> GitHubResult<FileCommitResponse> {
        self.client
            .put(&format!("/repos/{}/{}/contents/{}", owner, repo, path), request)
            .await
    }

    /// Deletes a file.
    pub async fn delete_file(
        &self,
        owner: &str,
        repo: &str,
        path: &str,
        request: &DeleteFileRequest,
    ) -> GitHubResult<FileCommitResponse> {
        // GitHub uses DELETE with a body for this operation
        self.client
            .raw_request(
                reqwest::Method::DELETE,
                &format!("/repos/{}/{}/contents/{}", owner, repo, path),
                Some(request),
            )
            .await?
            .json()
            .await
            .map_err(|e| {
                crate::errors::GitHubError::deserialization(format!(
                    "Failed to deserialize response: {}",
                    e
                ))
            })
    }

    /// Gets the README.
    pub async fn get_readme(&self, owner: &str, repo: &str) -> GitHubResult<Content> {
        self.client
            .get(&format!("/repos/{}/{}/readme", owner, repo))
            .await
    }

    // Releases

    /// Lists releases.
    pub async fn list_releases(&self, owner: &str, repo: &str) -> GitHubResult<Vec<Release>> {
        self.client
            .get(&format!("/repos/{}/{}/releases", owner, repo))
            .await
    }

    /// Gets a release.
    pub async fn get_release(&self, owner: &str, repo: &str, release_id: u64) -> GitHubResult<Release> {
        self.client
            .get(&format!("/repos/{}/{}/releases/{}", owner, repo, release_id))
            .await
    }

    /// Gets the latest release.
    pub async fn get_latest_release(&self, owner: &str, repo: &str) -> GitHubResult<Release> {
        self.client
            .get(&format!("/repos/{}/{}/releases/latest", owner, repo))
            .await
    }

    /// Gets a release by tag.
    pub async fn get_release_by_tag(
        &self,
        owner: &str,
        repo: &str,
        tag: &str,
    ) -> GitHubResult<Release> {
        self.client
            .get(&format!("/repos/{}/{}/releases/tags/{}", owner, repo, tag))
            .await
    }

    /// Creates a release.
    pub async fn create_release(
        &self,
        owner: &str,
        repo: &str,
        request: &CreateReleaseRequest,
    ) -> GitHubResult<Release> {
        self.client
            .post(&format!("/repos/{}/{}/releases", owner, repo), request)
            .await
    }

    /// Updates a release.
    pub async fn update_release(
        &self,
        owner: &str,
        repo: &str,
        release_id: u64,
        request: &UpdateReleaseRequest,
    ) -> GitHubResult<Release> {
        self.client
            .patch(
                &format!("/repos/{}/{}/releases/{}", owner, repo, release_id),
                request,
            )
            .await
    }

    /// Deletes a release.
    pub async fn delete_release(&self, owner: &str, repo: &str, release_id: u64) -> GitHubResult<()> {
        self.client
            .delete(&format!("/repos/{}/{}/releases/{}", owner, repo, release_id))
            .await
    }
}

/// Parameters for listing repositories.
#[derive(Debug, Clone, Default, Serialize)]
pub struct ListReposParams {
    /// Type filter.
    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
    pub repo_type: Option<RepoType>,
    /// Sort field.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sort: Option<RepoSort>,
    /// Sort direction.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub direction: Option<SortDirection>,
    /// Page number.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page: Option<u32>,
    /// Items per page.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub per_page: Option<u32>,
}

/// Repository type filter.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum RepoType {
    All,
    Owner,
    Public,
    Private,
    Member,
}

/// Repository sort field.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum RepoSort {
    Created,
    Updated,
    Pushed,
    FullName,
}

/// Sort direction.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum SortDirection {
    Asc,
    Desc,
}

/// Request to create a repository.
#[derive(Debug, Clone, Serialize)]
pub struct CreateRepoRequest {
    /// Repository name.
    pub name: String,
    /// Repository description.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Homepage URL.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub homepage: Option<String>,
    /// Whether the repository is private.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub private: Option<bool>,
    /// Whether issues are enabled.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub has_issues: Option<bool>,
    /// Whether projects are enabled.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub has_projects: Option<bool>,
    /// Whether wiki is enabled.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub has_wiki: Option<bool>,
    /// Auto-initialize with README.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auto_init: Option<bool>,
    /// Gitignore template.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gitignore_template: Option<String>,
    /// License template.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub license_template: Option<String>,
}

/// Request to update a repository.
#[derive(Debug, Clone, Default, Serialize)]
pub struct UpdateRepoRequest {
    /// Repository name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Repository description.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Homepage URL.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub homepage: Option<String>,
    /// Whether the repository is private.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub private: Option<bool>,
    /// Default branch.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_branch: Option<String>,
    /// Whether the repository is archived.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub archived: Option<bool>,
}

/// Request to create or update a file.
#[derive(Debug, Clone, Serialize)]
pub struct CreateOrUpdateFileRequest {
    /// Commit message.
    pub message: String,
    /// File content (base64 encoded).
    pub content: String,
    /// SHA of the file being replaced (for updates).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sha: Option<String>,
    /// Branch name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch: Option<String>,
    /// Committer information.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub committer: Option<CommitAuthor>,
    /// Author information.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub author: Option<CommitAuthor>,
}

/// Request to delete a file.
#[derive(Debug, Clone, Serialize)]
pub struct DeleteFileRequest {
    /// Commit message.
    pub message: String,
    /// SHA of the file being deleted.
    pub sha: String,
    /// Branch name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch: Option<String>,
    /// Committer information.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub committer: Option<CommitAuthor>,
    /// Author information.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub author: Option<CommitAuthor>,
}

/// Commit author information.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommitAuthor {
    /// Author name.
    pub name: String,
    /// Author email.
    pub email: String,
}

/// Response from file commit operations.
#[derive(Debug, Clone, Deserialize)]
pub struct FileCommitResponse {
    /// The committed content.
    pub content: Option<Content>,
    /// The commit.
    pub commit: FileCommit,
}

/// Commit information from file operations.
#[derive(Debug, Clone, Deserialize)]
pub struct FileCommit {
    /// Commit SHA.
    pub sha: String,
    /// Commit message.
    pub message: String,
    /// Commit URL.
    pub html_url: String,
}

/// Request to create a release.
#[derive(Debug, Clone, Serialize)]
pub struct CreateReleaseRequest {
    /// Tag name.
    pub tag_name: String,
    /// Target commitish.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_commitish: Option<String>,
    /// Release name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Release body.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<String>,
    /// Whether it's a draft.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub draft: Option<bool>,
    /// Whether it's a prerelease.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prerelease: Option<bool>,
    /// Generate release notes.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub generate_release_notes: Option<bool>,
}

/// Request to update a release.
#[derive(Debug, Clone, Default, Serialize)]
pub struct UpdateReleaseRequest {
    /// Tag name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tag_name: Option<String>,
    /// Target commitish.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_commitish: Option<String>,
    /// Release name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Release body.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<String>,
    /// Whether it's a draft.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub draft: Option<bool>,
    /// Whether it's a prerelease.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prerelease: Option<bool>,
}
